File: PL/Exc3/Prolog/ztalloc.pl
conditions(Dictionary,Mid,0):-
    get_assoc(Mid,Dictionary,Value), Value =< 0 -> true;false.
conditions(Dictionary,Mid,1):-
    get_assoc(Mid,Dictionary,Value), Value =< 1 -> true;false.
conditions(Dictionary,Mid,Ranger):-
    get_assoc(Mid,Dictionary,Value), Value =< Ranger -> true;
                                                        NewRanger is Ranger - 1, get_assoc(Mid-1,Dictionary,NewRanger),get_assoc(Mid+1,Dictionary,NewRanger).

foreveryinput([],Answer,X):-
    reverse(Answer,NewAnswer),
    X = NewAnswer.

foreveryinput([Head|Tail],Answer,X):-
    [Lin, Rin, _, _] = Head,
    empty_assoc(Dictionary),
    Mid is (Lin+Rin)//2, Range is Rin - Mid,
    put_assoc(Mid, Dictionary, Range, NewDictionary),
    %print(Tail), nl,
    solver(Head,[],[],NewDictionary,Answer,[],['h'],Tail,0,X).

success(Lin,Rin,Lout,Rout):-
    Lin >= Lout,
    Rin =< Rout.

solver(_,[],[],_,Answer,_,[],Tail,_,X):-
    NewAnswer = ['IMPOSSIBLE'|Answer],foreveryinput(Tail,NewAnswer,X).

solver(_,_,_,_,Answer,_,_,Tail,1000000,X):-
    NewAnswer = ['IMPOSSIBLE'|Answer],foreveryinput(Tail,NewAnswer,X).

solver([Lin,Rin,Lout,Rout],_,_,_,Answer,[],_,Tail,_,X):-
    success(Lin,Rin,Lout,Rout) -> NewAnswer = ['EMPTY'|Answer],foreveryinput(Tail,NewAnswer,X).

solver([_,_,Lout,Rout],[],[SecondH|SecondT],NewDictionary,Answer,_,[],Tail,I,X):-
    NewI is I+1,
    reverse([SecondH|SecondT],NewSecond),
    %print(NewSecond), nl,
    [SecH|SecT] = NewSecond,
    [String,Lin,Rin] = SecH,
    %length(String,Length),
    %print(Length),nl,%print(String),print(' '),print(Lin),print(' '),print(Rin),print(' '),print(Lout),print(' '),print(Rout),print(' '),nl,
    (success(Lin,Rin,Lout,Rout) -> reverse(String,Temp1),atomic_list_concat(Temp1,Temp2),NewAnswer = [Temp2|Answer],Temp = Tail,foreveryinput(Temp,NewAnswer,X);
                                solver([Lin,Rin,Lout,Rout],SecT,[],NewDictionary,Answer,String,['h'],Tail,NewI,X)).

solver([_,_,Lout,Rout],[FirstH|FirstT],Second,NewDictionary,Answer,_,[],Tail,I,X):-
    NewI is I+1,
    [String,Lin,Rin] = FirstH,
    %length(String,Length),
    %print(Length),nl,%print(String),print(' '),print(Lin),print(' '),print(Rin),print(' '),print(Lout),print(' '),print(Rout),print(' '), nl,
    (success(Lin,Rin,Lout,Rout) -> reverse(String,Temp1),atomic_list_concat(Temp1,Temp2),NewAnswer = [Temp2|Answer],Temp = Tail,foreveryinput(Temp,NewAnswer,X);
                                solver([Lin,Rin,Lout,Rout],FirstT,Second,NewDictionary,Answer,String,['h'],Tail,NewI,X)).

solver([Lin,Rin,Lout,Rout],First,[],NewDictionary,Answer,String,[],Tail,I,X):-
    solver([Lin,Rin,Lout,Rout],[],First,NewDictionary,Answer,String,[],Tail,I,X).

solver([Lin,Rin,Lout,Rout],First,Second,Dictionary,Answer,String,['h'],Tail,I,X):-
    L1 is Lin//2, R1 is Rin//2, Mid is (R1+L1)//2, Range is R1 - Mid,
    (conditions(Dictionary,Mid,Range) -> solver([Lin,Rin,Lout,Rout],First,Second,Dictionary,Answer,String,['t'],Tail,I,X);
        put_assoc(Mid,Dictionary,Range,NewDictionary),solver([Lin,Rin,Lout,Rout],First,[[['h'|String],L1,R1]|Second],NewDictionary,Answer,String,['t'],Tail,I,X)).

solver([Lin,Rin,Lout,Rout],First,Second,Dictionary,Answer,String,['t'],Tail,I,X):-
    L1 is (Lin*3)+1, R1 is (Rin*3)+1, Mid is (R1+L1)//2, Range is R1 - Mid,
    (R1 > 1000000 -> solver([Lin,Rin,Lout,Rout],First,Second,Dictionary,Answer,String,[],Tail,I,X);
    (conditions(Dictionary,Mid,Range)->  solver([Lin,Rin,Lout,Rout],First,Second,Dictionary,Answer,String,[],Tail,I,X);
         put_assoc(Mid,Dictionary,Range,NewDictionary),solver([Lin,Rin,Lout,Rout],First,[[['t'|String],L1,R1]|Second],NewDictionary,Answer,String,[],Tail,I,X))).

ztalloc(File,Answer):-
    open(File, read, Str),
    read_file(Str,[_|Lines]),
    close(Str),
    once(foreveryinput(Lines,[],Answer)).

read_file(Stream,[]) :-
    at_end_of_stream(Stream).

read_file(Stream,[X|L]) :-
    \+ at_end_of_stream(Stream),
    read_line_to_codes(Stream,Codes),
    atom_codes(Atom, Codes),
    atomic_list_concat(Atoms, ' ', Atom),
    maplist(atom_number, Atoms, X),
    read_file(Stream,L), !.

File: PL/Exc3/Prolog/Ztalloc.py
import sys
import collections
import time

def success(Lin,Rin,Lout,Rout):
    return (Lin>=Lout and Rout>=Rin)

def conditions(Set,mid,ranger):
    if ranger == 0 or ranger == 1:
        temp = Set.get(mid)
        if temp!= None and temp <= ranger:
            return False
        else:
            return True
    for i in range (0,ranger-1):
        temp = Set.get(mid-i)
        if temp!= None and temp<ranger-i:
            return False
        temp = Set.get(mid+i)
        if temp!= None and temp < ranger -i:
            return False
    return True

def solver2(answer,Lin,Rin,Lout,Rout):
    tries = collections.deque()
    sum = 0
    answer = list()
    Set = {}
    mid = (Lin+Rin)//2
    range = Rin - mid
    Set.update({mid:range})
    while success(Lin,Rin,Lout,Rout) == False:
        L1 = Lin//2
        R1 = Rin//2
        mid1 = (L1+R1)//2
        range1 = R1 - mid1
        if (conditions(Set,mid1,range1)):
            copy1 = list(answer)
            copy1.append('h')
            Set.update({mid1:range1})
            tries.append((copy1,L1,R1))
        L2 = Lin*3+1
        R2 = Rin*3+1
        mid2 = (L2+R2)//2
        range2 = R2 - mid2
        if R2 <= 1000000:
            if (conditions(Set,mid2,range2)):
                answer.append('t')
                Set.update({mid2:range2})
                tries.append((answer,L2,R2))
        if tries:
            temp = tries.popleft()
            sum = sum +1
            answer,Lin,Rin = temp[0],temp[1],temp[2]
        else:
            return "IMPOSSIBLE"
    if answer == []:
        return "EMPTY"
    else:
        return "".join(answer)

def solver1(answer,Lin,Rin,Lout,Rout):
    tries = collections.deque()
    sum = 0
    answer = list()
    Set = {}
    mid = (Lin+Rin)//2
    range = Rin - mid
    Set.update({mid:range})
    while success(Lin,Rin,Lout,Rout) == False:
        sum = sum +1
        L1 = Lin//2
        R1 = Rin//2
        mid1 = (L1+R1)//2
        #print(mid)
        range1 = R1 - mid1
        temp = Set.get(mid1)
        if range1 > 50:
            check1 = (temp == None or temp > range1)
        else:
            check1 = (conditions(Set,mid1,range1))
        if (check1):
            copy1 = list(answer)
            copy1.append('h')
            Set.update({mid1:range1})
            tries.append((copy1,L1,R1))
        L2 = Lin*3+1
        R2 = Rin*3+1
        mid2 = (L2+R2)//2
        range2 = R2 - mid2
        temp = Set.get(mid2)
        if R2 <= 1000000:
            if range2 > 50:
                check2 = (temp == None or temp > range2)
            else:
                check2 = (conditions(Set,mid2,range2))
            if (check2):
                answer.append('t')
                Set.update({mid2:range2})
                tries.append((answer,L2,R2))
        if tries:
            temp = tries.popleft()
            answer,Lin,Rin = temp[0],temp[1],temp[2]
        else:
            print(sum)
            return "IMPOSSIBLE"
    if answer == []:
        print(sum)
        return "EMPTY"
    else:
        print(sum)
        return "".join(answer)

def main():
    if len(sys.argv) < 2:
        print("No File Given!")
        sys.exit()
    filename = sys.argv[1]
    file = open(filename, "r")
    if file.mode == 'r':
        Q = file.readline()
        for i in file.readlines():
            Lin , Rin , Lout , Rout = i.split()
            Lin = int(Lin)
            Rin = int(Rin)
            Lout = int(Lout)
            Rout = int(Rout)
            #print(Lin,Rin,Lout,Rout)
            print(solver1("",Lin,Rin,Lout,Rout))
        return

if __name__ == '__main__':
    main()
